### PR TITLE
[New feature] Version History limit context

### PR DESCRIPTION
**Pull request by Devportobello**
## Context

When using Joomla! versioning feature, we are able to fix a number of history limit in configuration.
[Follow of #7191](https://github.com/joomla/joomla-cms/issues/7191)
## Actual limitation

This history limit is only applied of the whole component context eg: com_banners.
If you want to got (wanna keep using banners component for exemple) only 10 history limit for banner and 5 for client (dunno if this could be real) this is actually not possible.
### Part1
- Create 1 banner, save and edit multiple time (you need versioning history ON and history limitation setup)
- Then create 1 client, save and edit multiple time.
- Both got same history limitation
- Erase all
### Part2

Apply patch.

Go to `administrator\components\com_banners` and edit `config.xml` and add after

``` xml
<field
    name="history_limit"
    type="text"
    filter="integer"
    label="JGLOBAL_HISTORY_LIMIT_OPTIONS_LABEL"
    description="JGLOBAL_HISTORY_LIMIT_OPTIONS_DESC"
    default="5"
/>
```

this

``` xml
<field
    name="history_limit_client"
    type="text"
    filter="integer"
    label="COM_BANNERS_CLIENT_HISTORY_LIMIT_OPTIONS_LABEL"
    description="COM_BANNERS_CLIENT_HISTORY_LIMIT_OPTIONS_DESC"
    default="5"
/>
<field
    name="history_limit_banner"
    type="text"
    filter="integer"
    label="COM_BANNERS_BANNER_HISTORY_LIMIT_OPTIONS_LABEL"
    description="COM_BANNERS_BANNER_HISTORY_LIMIT_OPTIONS_DESC"
    default="5"
/>
```

OR
Use [this repo / ZIP](https://github.com/Devportobello/joomla-attach-PR-7206).

Repeat the part1 after having different setup for history limitation using 2 new context (banner / client) eg: 3 for maximum version client, 4 for maximum version banner and 5 for the base maximum version
You would be able to got different number of history limit.

### CODE REVIEW

**Comment by smz** [comment]
The link to "this repo / ZIP" mentioned in the PR description does not work.

Why don't you create a "sister PR" for com_banners (*which btw should be on its way out...*)? That would make testers life a lot easier...

If the com_banners mods is not meant to stay there you can clearly state that in the PR description and then close it when/if this PR is merged.

**Comment by zero-24** [comment]
@smz @Devportobello try this banner release: https://github.com/zero-24/banners/releases/tag/version_history_test

follow the instruction on the release and here. For me it works Thanks @Devportobello

**Comment by jwaisner** [comment]
@test 

PR works as intended.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/7206">issues.joomla.org/joomla-cms/7206</a>.</sub>

**Comment by zero-24** [comment]
RTC. Thanks :smile: <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/7206">issues.joomla.org/joomla-cms/7206</a>.</sub>

File: libraries/cms/helper/contenthistory.php [hunk 1]
```diff
@@ -145,7 +145,15 @@ public function store($table)
 		// Load history_limit config from extension.
 		$aliasParts = explode('.', $this->typeAlias);
 
-		if ($maxVersions = JComponentHelper::getParams($aliasParts[0])->get('history_limit', 0))
+		$context = (isset($aliasParts[1])) ? $aliasParts[1] : '';
+
+		$maxVersionsContext = JComponentHelper::getParams($aliasParts[0])->get('history_limit' . '_' . $context, null);
```
Comment (zero-24) [inline libraries/cms/helper/contenthistory.php]: @Devportobello just a codestyle change. I think we should use `0` here insted of `Null`

File: libraries/cms/helper/contenthistory.php [hunk 2]
```diff
@@ -145,7 +145,15 @@ public function store($table)
 		// Load history_limit config from extension.
 		$aliasParts = explode('.', $this->typeAlias);
 
-		if ($maxVersions = JComponentHelper::getParams($aliasParts[0])->get('history_limit', 0))
+		$context = (isset($aliasParts[1])) ? $aliasParts[1] : '';
+
+		$maxVersionsContext = JComponentHelper::getParams($aliasParts[0])->get('history_limit' . '_' . $context, 0);
```
Comment (roland-d) [inline libraries/cms/helper/contenthistory.php]: I think we have an issue here. If the $context isn't set, you are trying to get the parameters from history_limit_ and that will never exist. The context check should be extend so you either check history_limit or a context-aware limit.

Comment (Devportobello) [inline libraries/cms/helper/contenthistory.php]: in that case, $maxVersionsContext gonna be like empty, so will check for $maxVersions (like old way) see #156

**Comment by roland-d** [comment]
Thanks everybody. Merged into 3.5-dev with commit ed82a74

**Pull request closed without merging**